### PR TITLE
Schedule tasks from the Task API

### LINKED ISSUE #9834: Enhance the tasks API endpoints to allow scheduling tasks and retreival of scheduled tasks
Implement:

- Tasks Create API endpoint allows scheduling tasks.
- Tasks List API endpoint allows retreival of only scheduled tasks.

These use cases came up when discussing with Allan about https://github.com/learningequality/kolibri/issues/9804.

cc @rtibbles @AllanOXDi

### PR DESCRIPTION
**Pull request by vkWeb**
## Summary

This PR enables scheduling of tasks in future from the frontend by use of Tasks API POST endpoint. Also, enables listing of tasks depending on whether they are repeating or not.

## References

Closes https://github.com/learningequality/kolibri/issues/9834.

## Reviewer guidance
Are all tests passing? How does the request structure looks? Is it user friendly?
----

## Testing checklist

- [x] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [x] Critical and brittle code paths are covered by unit tests

## PR process

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

File: kolibri/core/tasks/storage.py [hunk 1]
```diff
@@ -524,6 +529,10 @@ def schedule(
         """
         if not isinstance(dt, datetime):
             raise ValueError("Time argument must be a datetime object.")
+        if repeat != None and repeat < 0:  # noqa E711
```
Comment (rtibbles) [inline kolibri/core/tasks/storage.py]: In this case you should use:
```suggestion
        if repeat is not None and repeat < 0:
```
We're only forced to use `==` and `!=` for `None` when inside a SQLAlchemy query.

File: kolibri/core/tasks/api.py [hunk 2]
```diff
@@ -56,14 +56,22 @@ def validate_create_req_data(self, request):
...
         for request_data in request_data_list:
-            # Make sure the task is registered
+            # Make sure the task is registered.
             registered_task = TaskRegistry.validate_task(request_data.get("type"))
 
+            # Make sure the job's data is valid along with `enqueue_args`.
             job = registered_task.validate_job_data(request.user, request_data)
 
+            # Get validated `enqueue_args`.
+            enqueue_args = registered_task.get_validated_enqueue_args(
```
Comment (rtibbles) [inline kolibri/core/tasks/api.py]: Not sure why this needs to be validated again? It should already have been validated by the job data validation.

File: kolibri/core/tasks/registry.py [hunk 3]
```diff
@@ -243,6 +245,15 @@ def check_job_permissions(self, user, job, view):
             if not permission.has_permission(user, job, view):
                 raise PermissionDenied
 
+    def get_validated_enqueue_args(self, enqueue_args_data):
+        """
+        The `JobValidator` pops out `enqueue_args` during validation so
+        this method serves the purpose of returning validated `enqueue_args`.
+        """
+        enqueue_args_serializer = EnqueueArgsSerializer(data=enqueue_args_data)
+        enqueue_args_serializer.is_valid(raise_exception=True)
```
Comment (rtibbles) [inline kolibri/core/tasks/registry.py:252]: This should already have been validated as part of a the parent serializer validation.

File: kolibri/core/tasks/validation.py [hunk 4]
```diff
@@ -9,10 +48,12 @@ class JobValidator(serializers.Serializer):
     """
 
     type = serializers.CharField(required=True)
+    enqueue_args = EnqueueArgsSerializer(required=False)
 
     def validate(self, data):
         kwargs = data.copy()
         kwargs.pop("type")
+        kwargs.pop("enqueue_args", None)
```
Comment (rtibbles) [inline kolibri/core/tasks/validation.py:56]: This should go on the returned dict.

**Review by rtibbles — COMMENTED** [review]
Would be good to use the single validation pass here. A couple of other comments.

File: kolibri/core/tasks/storage.py (hunk 1 above)
Comment (vkWeb) [reply kolibri/core/tasks/storage.py]: Fixed sir.

**Comment by vkWeb** [comment]
@rtibbles sir review comments are addressed as discussed on Slack sir. This is ready to go sir.

File: kolibri/core/tasks/test/test_api.py [hunk 5]
```diff
@@ -520,6 +535,308 @@ def add(x, y):
...
+
+        datetime_obj = datetime.datetime(year=2023, month=1, day=1, tzinfo=pytz.utc)
+        timedelta_obj = datetime.timedelta(days=1, hours=1)
+
+        cls.enqueue_at_datetime = str(datetime_obj)
+        cls.enqueue_in_timedelta = str(timedelta_obj)
+
+    def setUp(self):
+        @register_task(job_id="test-id")
+        def life():
```
Comment (rtibbles) [inline kolibri/core/tasks/test/test_api.py:551]: This function appears to be misnamed - it should be `life_the_universe_and_everything` surely?

File: kolibri/core/tasks/validation.py [hunk 6]
```diff
@@ -1,5 +1,44 @@
...
+from kolibri.core.serializers import DateTimeTzField
+
+
+class EnqueueArgsSerializer(serializers.Serializer):
+    """
+    A serializer for `enqueue_args` object of incoming user request data.
+    """
+
+    enqueue_at = DateTimeTzField(required=False)
+    enqueue_in = serializers.DurationField(required=False)
```
Comment (rtibbles) [inline kolibri/core/tasks/validation.py:12]: Just realizing that we will need to have a utility function on the frontend to produce this duration value in the form `'[DD] [HH:[MM:]]ss[.uuuuuu]'`.

**Review by rtibbles — APPROVED** [review]

**Comment by rtibbles** [comment]
Relevant tests pass locally, so merging this. I'll clean up any resultant issues.

File: kolibri/core/tasks/test/test_api.py (hunk 5 above)
Comment (vkWeb) [reply kolibri/core/tasks/test/test_api.py:551]: 🤣 Yeah that's definitely more complete name!

**Pull request merged**